### ISSUE TITLE
Impossible to use mouse to move cursor to the left

### ISSUE THREAD

**Issue by ssh-peppe** [xtermjs/xterm.js#2639]
In certain situations, such as when in a nano text editing session, it seems like it's impossible to move the cursor to the left using the mouse. Alt-clicking the mouse moves the cursor as expected as long as the direction is up or down or to the right. But trying to move it to the left simply does nothing.

Additionally, if I keep clicking the terminal goes into some kind of busy loop that eventually results in the browser or application (also tested in Visual Studio Code) to crash or terminate the script.

I know there is an issue with UTF-8 encodings and mouse cursor commands not being compatible (resulting in invalid UTF-8 sequences) but even though I'm also registering an onBinary listener it seems it never gets called. It appears that onData is getting called with a zero length string when trying to move the cursor to the left and that fails.

## Details
- Browser and browser version: All tested browsers (Safari, Firefox, Chrome) as well as MS Visual Studio Code
- OS version: macOS 10.15.2
- xterm.js version: 4.3.0, as well as whatever ships with the latest VS Code

### Steps to reproduce

1a. Connect to a server and start the nano editor
1b. Alternatively start a terminal in VS Code and start the nano editor
2. Write few short lines of text
3. Move the cursor to the left of a line
4. alt-click to move the cursor towards the right
5. The bug is that if you now alt-click to move the cursor towards the left, it doesn't in fact move
6. Keep alt-clicking around on the text that is on the left of the current cursor position and eventually (after maybe 10-30 seconds) the app (browser or VS Code) becomes unresponsive

Not sure if this is also a related bug or not, but the onBinary() event handler never fires. The documentation is not clear about whether it should fire in addition to onData() or only in some situations when onData() does not fire. A quick glance at the xterm.js code suggests its always either onData or onBinary - never both, but I'm not 100% sure about this.

**Comment by mofux** [comment]
I'm able to reproduce this, also on macOS against the vscode terminal.  The cursor move via `alt+click` is implemented on our side, I expect the bug to be found somwhere along these lines:

https://github.com/xtermjs/xterm.js/blob/71a78cbf0159adab95aaeb6e6a11cfa7485aca9c/src/browser/input/MoveToCell.ts#L183-L200

**Comment by jerch** [comment]
Oh I wasnt aware we have that feature, always thought Alt+Click is only about rect selections. How is this supposed to behave and under with circumstances? I see in the logs that this sends arrow key sequences to the backend. Here we have a problem: If the backend does not deal with arrow keys at all, this will produce really ugly nonsense output (like in termios ICANON mode), thus I am not sure if we should try to mess with the cursor at all this way?

@ssh-peppe The onBinary event is not coupled to this, it is only used by nano, if you enabled mouse support in nano by pressing Alt+m. With mouse support in nano you can simply click where you want the cursor.

**Comment by mofux** [comment]
@jerch Some context can be found in #896

**Comment by jerch** [comment]
@mofux Thx. Hmm still unsure about this, to me it feels like a hack to achieve something that is likely not to work for several reasons:
- arrow keys are not always coupled to cursor movements by the app (it just happens to work with some that way, others like libreadline bind history lookup to arrow up - urgh)
- cursor responsibility in on app side (with some predefined terminal contraints like wrapping)
- cursor repositioning is a "sensitive" topic for almost any app, thus cannot be achieved by a "one fits all" from terminal side

Given that an app alone should decide whether it wants to offer free repositioning to the user and should register mouse reporting and deal with it. It knows its own cursor restrictions, thus can deal with all its state edge cases. Our terminal driven hack here never can do that?

**Comment by ssh-peppe** [comment]
I'd like to remind again that while alt-clicking around to the left side of the cursor position in nano, I can get both VS Code and my own app that uses xterm.js to completely freeze up (spinning beach ball on macOS). So definitely seems like something in the code isn't working right.

**Comment by jerch** [comment]
@ssh-peppe Yes this is serious, I think it can even happen accidently.  So thx for finding this one. @mofux already pointed to some fishy code (a while loop messing with conditional vars into both directions is always suspicous to get reduced to `while (true) {}`).

**Comment by mofux** [comment]
Back in those days I guess we thought: `Terminal.app` does it, so why shouldn't we. 

Nowadays I would propose to get rid of that feature in xterm.js core and maybe move it to an addon (if the freeze bug can be fixed). Personally I'm not interested in this feature. And as @jerch pointed out, it's not a good solution anyway.

@Tyriar @npezza93 Any thoughts?

**Comment by Tyriar** [comment]
Moving this into an addon seems like a good move, would need to be included with a new major version though.

**Comment by davidkhala** [comment]
BTW, arrow key problem also apear in Xterm on GRUB boot menu edit mode, which can't move the cursor anyway
